django/contrib/auth/forms.py

```diff
@@ -89,50 +89,108 @@ def widget_attrs(self, widget):
+        label=_("Password-based authentication"),
+        required=False,
+        initial="true",
+        choices={"true": _("Enabled"), "false": _("Disabled")},
+        widget=forms.RadioSelect(
+            attrs={"class": "radiolist inline"},
+        ),
+        help_text=usable_help_text,
+    )
     password1 = forms.CharField(
         label=_("Password"),
         strip=False,
+        required=False,
         widget=forms.PasswordInput(attrs={"autocomplete": "new-password"}),
         help_text=password_validation.password_validators_help_text_html(),
     )
     password2 = forms.CharField(
         label=_("Password confirmation"),
         widget=forms.PasswordInput(attrs={"autocomplete": "new-password"}),
         strip=False,
+        required=False,
         help_text=_("Enter the same password as before, for verification."),
     )
 
-    class Meta:
-        model = User
-        fields = ("username",)
-        field_classes = {"username": UsernameField}
-
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args, extra_warning=None, **kwargs):
         super().__init__(*args, **kwargs)
-        if self._meta.model.USERNAME_FIELD in self.fields:
-            self.fields[self._meta.model.USERNAME_FIELD].widget.attrs[
-                "autofocus"
-            ] = True
+        if extra_warning:
+            self.fields["usable_password"].help_text += (
+                f'<ul id="id_unusable_warning" class="messagelist">'
+                f'<li class="warning">{extra_warning}</li></ul>'
+            )
+
+    def clean_password1(self):
+        password1 = self.cleaned_data.get("password1")
+        if self.data.get("usable_password") == "true" and not password1:
+            raise ValidationError(
+                self.fields["password1"].error_messages["required"], code="required"
+            )
+        return password1
 
     def clean_password2(self):
         password1 = self.cleaned_data.get("password1")
         password2 = self.cleaned_data.get("password2")
+        if self.data.get("usable_password") == "false":
+            return password2
+        if not password2:
+            raise ValidationError(
+                self.fields["password2"].error_messages["required"], code="required"
+            )
         if password1 and password2 and password1 != password2:
             raise ValidationError(
                 self.error_messages["password_mismatch"],
                 code="password_mismatch",
             )
         return password2
 
+    def set_password_and_save(self, user, commit=True):
+        if self.cleaned_data["usable_password"] == "true":
+            user.set_password(self.cleaned_data["password1"])
+        else:
+            user.set_unusable_password()
+        if commit:
+            user.save()
+            if hasattr(self, "save_m2m"):
+                self.save_m2m()
+        return user
+
+
+class BaseUserCreationForm(SetPasswordMixin, forms.ModelForm):
```

Comment:
This class is [documented](https://docs.djangoproject.com/en/dev/topics/auth/default/#django.contrib.auth.forms.BaseUserCreationForm), and its documentation no longer matches the implementation.

Reply:
@felixxm Thanks for pointing this out. I've updated the documentation.